### ISSUE TITLE
`E490: no fold found` with tree-sitter

### ISSUE THREAD

**Issue by adrian5** [neovim/neovim#28692]
### Problem

Tree-sitter computed folds don't work on files opened through commands. This breaks session restore functionality.

Related:

* #25948
* [nvim-treesitter #1337](https://github.com/nvim-treesitter/nvim-treesitter/issues/1337)
* [nvim-telescope #699](https://github.com/nvim-telescope/telescope.nvim/issues/699)
* [auto-session #109](rmagatti/auto-session/issues/109)
* [possession.nvim #19](https://github.com/jedrzejboczar/possession.nvim/issues/19)

### Steps to reproduce

**minimal.lua**:
```lua
for name, url in pairs {
  ts = "https://github.com/nvim-treesitter/nvim-treesitter/",
} do
  local install_path = vim.fn.fnamemodify('nvim_issue/' .. name, ':p')
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
  vim.opt.runtimepath:append(install_path)
end

require("nvim-treesitter.configs").setup({
  ensure_installed = { "c" }
})

vim.opt.foldmethod = "expr"
vim.wo.foldexpr = "v:lua.vim.treesitter.foldexpr()"
vim.opt.foldlevelstart = 99

vim.cmd("e hello-world.c")
vim.cmd("setf c")
```
**helloworld.c**:
```c
#include <stdio.h>

int main() {
    printf("Hello, world!");
    return 0;
}
```

**Steps**:

* `nvim --clean -u minimal.lua`
* Press `za` on foldable section
* message: `E490: no fold found`

Entering insert mode or pressing `zx` seems to restore fold functionality. Not working without these actions breaks a complete session restore.

### Expected behavior

Folds should always work, including those restored via sessions.

### Neovim version (nvim -v)

NVIM v0.10.0-dev-3103+gca735c755

### Vim (not Nvim) behaves the same?

not applicable

### Operating system/version

Arch Linux x86_64

### Terminal name/version

Alacritty 0.13.2-2

### $TERM environment variable

alacritty

### Installation

GitHub nightly tarball

**Comment by clason** [comment]
You need to set these options in a FileType autocommand or ftplugin.

**Comment by adrian5** [comment]
And should `:mksession` work with options the way you suggest? From the help file it looks like it might only track global options.

If it does not restore local options, does that mean I cannot restore folds (`sessionoptions`) in conjunction with nvim-treesitter?

**Comment by clason** [comment]
The folds from treesitter behave exactly like other folds.

**Comment by adrian5** [comment]
As sparse as ever with your answers. I would have liked to figure this out with vanilla nvim, but will return to using [nvim-ufo](https://github.com/kevinhwang91/nvim-ufo/) then, which somehow handles these use cases.

**Comment by clason** [comment]
That's all the time I have for, unfortunately. You can try your luck with @lewis6991.

**Comment by adrian5** [comment]
Fair enough, just my frustration speaking as I've gotten stuck on this far too long, apologies!

**Comment by wookayin** [comment]
Try evaluating `vim.treesitter.foldexpr()` manually, e.g. `:lua= vim.treesitter.foldexpr(3)` and see what it prints. Check `setlocal foldexpr` and `setlocal foldmethod` too.

Also ensure the TS parser exists and works for the buffer: `:lua= vim.treesitter.get_parser(0):lang()`.

**Comment by tomtomjhj** [comment]
> And should :mksession work with options set the way you suggest? 

Yes that works. The problem in the repro script is that filetype is not set when foldexpr is first evaluated, so the foldexpr doesn't recognize the language. Setting up an FileType autocmd to set foldexpr will reevaluate foldexpr with proper filetype.

**Comment by asilvadesigns** [comment]
we bit of a workaround but it helped me...

```
vim.schedule(function()
  vim.opt.foldmethod = "expr"
  vim.opt.foldexpr = "v:lua.vim.treesitter.foldexpr()"
end)
```

**Comment by faergeek** [comment]
The issue with session restoration seems to be caused by treesitter parsing becoming async. This fixed it for me even though I just set options through `vim.opt` outside of any autocommand:

```lua
vim.g._ts_force_sync_parsing = true
```

In the session file fold states are saved as a list of commands, something like this:

```vim
14 " go to line 14
normal! zo " open the fold
24 " go to line 24
normal! zc " close the fold
```

And if there is no fold on that line yet (because of async parsing) it will just error out.

Personally, I ended up not adding `folds` to `sessionoptions` as they vanish after a simple `:edit`. Not sure if I might be doing something wrong or that's just the way it is, didn't find about it in the docs. In practice I mostly only use them in diff mode anyway.

**Comment by igorlfs** [comment]
> The issue with session restoration seems to be caused by treesitter parsing becoming async

No, not really. This issue was first reported way before async TS.

---

With https://github.com/vim/vim/pull/17127, the error is no longer thrown when restoring sessions. In fact, from my testing, folds are always restored properly (would be nice if someone could confirm).
